### PR TITLE
Refactor `modules lint`

### PR DESCRIPTION
**Pull request by ErikDanielsson**
Another try at refactoring `nf-core modules lint` and adding module tests to `nf-core lint`: 
- Moved code into methods and removed duplicated methods.
- Added modules tests to `nf-core lint` and made the general lint table and the module lint table work together.
- Created `lint_utils.py` for code used by both `modules lint` and `lint` commands.

On the way to resolving #1158. If these changes make sense, we should also update the markdown and json generation by `nf-core lint` to also include the modules tests. 

## PR checklist

 - [x] This comment contains a description of changes (with reason)
 - [ ] `CHANGELOG.md` is updated
 - [ ] If you've fixed a bug or added code that should be tested, add tests!
 - [ ] Documentation in `docs` is updated

### CODE REVIEW

File: nf_core/modules/lint/__init__.py [hunk 1]
```diff
@@ -148,121 +157,72 @@ def lint(self, module=None, all_modules=False, print_results=True, show_passed=F
...
-                )
-            )
-
-        # If -k supplied, only run these tests
-        if self.key:
-            log.info("Only running tests: '{}'".format("', '".join(self.key)))
-            self.lint_tests = [k for k in self.lint_tests if k in self.key]
+        # Filter the tests by the key if one is supplied
+        if key:
+            self.filter_tests_by_key()
```
Comment (KevinMenden) [inline nf_core/modules/lint/__init__.py:162]: Don't we need to supply the `key` to the filter function here?

Comment (ErikDanielsson) [reply nf_core/modules/lint/__init__.py:162]: You're right. Went back and forth between having `key` as a argument and as an attribute of the class.

File: nf_core/modules/lint/__init__.py [hunk 2]
```diff
@@ -148,121 +157,72 @@ def lint(self, module=None, all_modules=False, print_results=True, show_passed=F
...
+        bad_keys = [k for k in key if k not in self.lint_tests]
+        if len(bad_keys) > 0:
+            raise AssertionError(
+                "Test name{} not recognised: '{}'".format(
+                    "s" if len(bad_keys) > 1 else "",
+                    "', '".join(bad_keys),
                 )
-                mod_object.main_nf = mod
-                mod_object.module_name = os.path.basename(mod)
-                self.lint_module(mod_object, local=True)
```
Comment (KevinMenden) [inline nf_core/modules/lint/__init__.py:214]: This is only looking for bad keys right?

Comment (KevinMenden) [reply nf_core/modules/lint/__init__.py:214]: ah nevermind, a bit downstream ...

File: nf_core/modules/lint/__init__.py [hunk 3]
```diff
@@ -288,15 +248,86 @@ def get_installed_modules(self):
...
+            for mod in modules:
+                progress_bar.update(lint_progress, advance=1, test_name=mod.module_name)
+                self.lint_module(mod, local=local)
+
+    def lint_module(self, mod, local=False):
+        """
+        Perform linting on one module
+
+        If the module is a local module we only check the `main.nf` file,
+        and it issue warnings instead of failures.
```
Comment (KevinMenden) [inline nf_core/modules/lint/__init__.py:303]:
```suggestion
        and issue warnings instead of failures.
```

**Review by KevinMenden — APPROVED** [review]
Looks good to me and works locally when testing! 👍 

Any further refactoring can still be handled after the v2.0 release then.

**Pull request merged**